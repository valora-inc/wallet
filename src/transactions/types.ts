import BigNumber from 'bignumber.js'
import { Nft } from 'src/nfts/types'
import { v4 as uuidv4 } from 'uuid'

export enum Network {
  Celo = 'celo',
  Ethereum = 'ethereum',
  Arbitrum = 'arbitrum',
  Optimism = 'optimism',
}

export enum NetworkId {
  'celo-mainnet' = 'celo-mainnet',
  'celo-alfajores' = 'celo-alfajores',
  'ethereum-mainnet' = 'ethereum-mainnet',
  'ethereum-sepolia' = 'ethereum-sepolia',
  'arbitrum-one' = 'arbitrum-one',
  'arbitrum-sepolia' = 'arbitrum-sepolia',
  'op-mainnet' = 'op-mainnet',
  'op-sepolia' = 'op-sepolia',
}

export type PendingStandbySwap = {
  transactionHash?: string
  context: TransactionContext
  status: TransactionStatus.Pending
  feeCurrencyId?: string
} & Omit<TokenExchange, 'block' | 'fees' | 'transactionHash' | 'status'>

export type PendingStandbyTransfer = {
  transactionHash?: string
  context: TransactionContext
  status: TransactionStatus.Pending
  feeCurrencyId?: string
} & Omit<TokenTransfer, 'block' | 'fees' | 'transactionHash' | 'status'>

export type PendingStandbyApproval = {
  transactionHash?: string
  context: TransactionContext
  status: TransactionStatus.Pending
  feeCurrencyId?: string
} & Omit<TokenApproval, 'block' | 'fees' | 'transactionHash' | 'status'>

export type ConfirmedStandbyTransaction = (
  | Omit<TokenExchange, 'status'>
  | Omit<TokenTransfer, 'status'>
  | Omit<TokenApproval, 'status'>
) & {
  status: TransactionStatus.Complete | TransactionStatus.Failed
  context: TransactionContext
  feeCurrencyId?: string
}

export type StandbyTransaction =
  | PendingStandbySwap
  | PendingStandbyTransfer
  | PendingStandbyApproval
  | ConfirmedStandbyTransaction

// Context used for logging the transaction execution flow.
export interface TransactionContext {
  // Unique identifier used for tracking a transaction within logging.
  // Note that this is not the transaction hash, which is unknown when creating a new transaction.
  id: string

  // A tag provided by the caller to provide context on the purpose.
  tag?: string

  // A short contextual description of what the transaction does. (e.g. "Approve attestations")
  description?: string
}

export function newTransactionContext(tag: string, description: string) {
  return {
    id: uuidv4(),
    tag,
    description,
  }
}

export interface PageInfo {
  startCursor: string
  endCursor: string
  hasNextPage: boolean
  hasPreviousPage: boolean
}

export enum TransactionStatus {
  Pending = 'Pending',
  Complete = 'Complete',
  Failed = 'Failed',
}

export type TokenTransaction = TokenTransfer | TokenExchange | NftTransfer | TokenApproval

export interface TokenAmount {
  value: BigNumber.Value
  tokenAddress?: string
  localAmount?: LocalAmount
  tokenId: string
}

export interface LocalAmount {
  value: BigNumber.Value
  currencyCode: string
  exchangeRate: string
}

export enum TokenTransactionTypeV2 {
  Exchange = 'EXCHANGE',
  Received = 'RECEIVED',
  Sent = 'SENT',
  InviteSent = 'INVITE_SENT',
  InviteReceived = 'INVITE_RECEIVED',
  NftReceived = 'NFT_RECEIVED',
  NftSent = 'NFT_SENT',
  SwapTransaction = 'SWAP_TRANSACTION',
  Approval = 'APPROVAL',
}

// Can we optional the fields `transactionHash` and `block`?
export interface TokenTransfer {
  __typename: 'TokenTransferV3'
  networkId: NetworkId
  type: TokenTransactionTypeV2
  transactionHash: string
  timestamp: number
  block: string
  address: string
  amount: TokenAmount
  metadata: TokenTransferMetadata
  fees: Fee[]
  status: TransactionStatus
}

export interface TokenTransferMetadata {
  title?: string
  subtitle?: string
  image?: string
  comment?: string
}

export interface NftTransfer {
  __typename: 'NftTransferV3'
  networkId: NetworkId
  type: TokenTransactionTypeV2
  transactionHash: string
  timestamp: number
  block: string
  fees?: Fee[]
  nfts?: Nft[]
  status: TransactionStatus
}

// Can we optional the fields `transactionHash` and `block`?
export interface TokenExchange {
  __typename: 'TokenExchangeV3'
  networkId: NetworkId
  type: TokenTransactionTypeV2
  transactionHash: string
  timestamp: number
  block: string
  inAmount: TokenAmount
  outAmount: TokenAmount
  metadata?: TokenExchangeMetadata
  fees: Fee[]
  status: TransactionStatus
}

export interface TokenExchangeMetadata {
  title?: string
  subtitle?: string
}

export enum FeeType {
  SecurityFee = 'SECURITY_FEE',
  GatewayFee = 'GATEWAY_FEE',
  EncryptionFee = 'ONE_TIME_ENCRYPTION_FEE',
}

export interface Fee {
  type: string
  amount: TokenAmount
}

export interface TokenApproval {
  __typename: 'TokenApproval'
  networkId: NetworkId
  type: TokenTransactionTypeV2
  timestamp: number
  block: string
  transactionHash: string
  tokenId: string
  approvedAmount: string | null // null represents infinite approval
  fees: Fee[]
  status: TransactionStatus
}
