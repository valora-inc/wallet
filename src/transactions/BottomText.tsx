import React from 'react'
import { StyleSheet, Text } from 'react-native'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'

interface BottomTextProps {
  children: React.ReactNode
}

export default function BottomText({ children }: BottomTextProps) {
  return <Text style={styles.text}>{children}</Text>
}

const styles = StyleSheet.create({
  text: {
    ...fontStyles.regular,
    color: colors.gray4,
    textAlign: 'center',
    marginTop: 'auto',
  },
})
