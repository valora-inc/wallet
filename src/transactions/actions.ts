import { CeloTxReceipt } from '@celo/connect'
import { SendOrigin } from 'src/analytics/types'
import { TokenTransactionType, TransactionFeedFragment } from 'src/apollo/types'
import { ExchangeConfirmationCardProps } from 'src/exchange/ExchangeConfirmationCard'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { NumberToRecipient } from 'src/recipients/recipient'
import { TransactionDataInput } from 'src/send/SendConfirmationLegacy'
import { InviteTransactions } from 'src/transactions/reducer'
import { TransferConfirmationCardProps } from 'src/transactions/TransferConfirmationCard'
import {
  StandbyTransaction,
  StandbyTransactionLegacy,
  TokenTransaction,
} from 'src/transactions/types'

export enum Actions {
  ADD_STANDBY_TRANSACTION = 'TRANSACTIONS/ADD_STANDBY_TRANSACTION',
  REMOVE_STANDBY_TRANSACTION = 'TRANSACTIONS/REMOVE_STANDBY_TRANSACTION',
  RESET_STANDBY_TRANSACTIONS = 'TRANSACTIONS/RESET_STANDBY_TRANSACTIONS',
  ADD_HASH_TO_STANDBY_TRANSACTIONS = 'TRANSACTIONS/ADD_HASH_TO_STANDBY_TRANSACTIONS',
  TRANSACTION_CONFIRMED = 'TRANSACTIONS/TRANSACTION_CONFIRMED',
  TRANSACTION_FAILED = 'TRANSACTIONS/TRANSACTION_FAILED',
  NEW_TRANSACTIONS_IN_FEED = 'TRANSACTIONS/NEW_TRANSACTIONS_IN_FEED',
  REFRESH_RECENT_TX_RECIPIENTS = 'TRANSACTIONS/REFRESH_RECENT_TX_RECIPIENTS',
  UPDATE_RECENT_TX_RECIPIENT_CACHE = 'TRANSACTIONS/UPDATE_RECENT_TX_RECIPIENT_CACHE',
  UPDATE_TRANSACTIONS = 'TRANSACTIONS/UPDATE_TRANSACTIONS',
  // Remove legacy action once the multitoken support feature is fully released
  ADD_STANDBY_TRANSACTION_LEGACY = 'TRANSACTIONS/ADD_STANDBY_TRANSACTION_LEGACY',
  UPDATE_INVITE_TRANSACTIONS = 'TRANSACTIONS/UPDATE_INVITE_TRANSACTIONS',
}

export interface AddStandbyTransactionAction {
  type: Actions.ADD_STANDBY_TRANSACTION
  transaction: StandbyTransaction
}

export interface RemoveStandbyTransactionAction {
  type: Actions.REMOVE_STANDBY_TRANSACTION
  idx: string
}

export interface ResetStandbyTransactionsAction {
  type: Actions.RESET_STANDBY_TRANSACTIONS
}

export interface AddStandbyTransactionLegacyAction {
  type: Actions.ADD_STANDBY_TRANSACTION_LEGACY
  transactionLegacy: StandbyTransactionLegacy
}

export interface AddHashToStandbyTransactionAction {
  type: Actions.ADD_HASH_TO_STANDBY_TRANSACTIONS
  idx: string
  hash: string
}

export interface TransactionConfirmedAction {
  type: Actions.TRANSACTION_CONFIRMED
  txId: string
  receipt: CeloTxReceipt
}

export interface TransactionFailedAction {
  type: Actions.TRANSACTION_FAILED
  txId: string
}

export interface NewTransactionsInFeedAction {
  type: Actions.NEW_TRANSACTIONS_IN_FEED
  transactions: TransactionFeedFragment[]
}

export interface UpdatedRecentTxRecipientsCacheAction {
  type: Actions.UPDATE_RECENT_TX_RECIPIENT_CACHE
  recentTxRecipientsCache: NumberToRecipient
}

export interface UpdateTransactionsAction {
  type: Actions.UPDATE_TRANSACTIONS
  transactions: TokenTransaction[]
}

export interface UpdateInviteTransactionsAction {
  type: Actions.UPDATE_INVITE_TRANSACTIONS
  inviteTransactions: InviteTransactions
}

export type ActionTypes =
  | AddStandbyTransactionAction
  | RemoveStandbyTransactionAction
  | ResetStandbyTransactionsAction
  | AddStandbyTransactionLegacyAction
  | AddHashToStandbyTransactionAction
  | NewTransactionsInFeedAction
  | UpdatedRecentTxRecipientsCacheAction
  | UpdateTransactionsAction
  | TransactionConfirmedAction
  | UpdateInviteTransactionsAction

export const addStandbyTransactionLegacy = (
  transaction: StandbyTransactionLegacy
): AddStandbyTransactionLegacyAction => ({
  type: Actions.ADD_STANDBY_TRANSACTION_LEGACY,
  transactionLegacy: transaction,
})

export const addStandbyTransaction = (
  transaction: StandbyTransaction
): AddStandbyTransactionAction => ({
  type: Actions.ADD_STANDBY_TRANSACTION,
  transaction,
})

export const removeStandbyTransaction = (idx: string): RemoveStandbyTransactionAction => ({
  type: Actions.REMOVE_STANDBY_TRANSACTION,
  idx,
})

export const updateRecentTxRecipientsCache = (
  recentTxRecipientsCache: NumberToRecipient
): UpdatedRecentTxRecipientsCacheAction => ({
  type: Actions.UPDATE_RECENT_TX_RECIPIENT_CACHE,
  recentTxRecipientsCache,
})

export const resetStandbyTransactions = (): ResetStandbyTransactionsAction => ({
  type: Actions.RESET_STANDBY_TRANSACTIONS,
})

export const transactionConfirmed = (
  txId: string,
  receipt: CeloTxReceipt
): TransactionConfirmedAction => ({
  type: Actions.TRANSACTION_CONFIRMED,
  txId,
  receipt,
})

export const transactionFailed = (txId: string): TransactionFailedAction => ({
  type: Actions.TRANSACTION_FAILED,
  txId,
})

export const addHashToStandbyTransaction = (
  idx: string,
  hash: string
): AddHashToStandbyTransactionAction => ({
  type: Actions.ADD_HASH_TO_STANDBY_TRANSACTIONS,
  idx,
  hash,
})

export const newTransactionsInFeed = (
  transactions: TransactionFeedFragment[]
): NewTransactionsInFeedAction => ({
  type: Actions.NEW_TRANSACTIONS_IN_FEED,
  transactions,
})

export const updateTransactions = (transactions: TokenTransaction[]): UpdateTransactionsAction => ({
  type: Actions.UPDATE_TRANSACTIONS,
  transactions,
})

export const updateInviteTransactions = (
  inviteTransactions: InviteTransactions
): UpdateInviteTransactionsAction => ({
  type: Actions.UPDATE_INVITE_TRANSACTIONS,
  inviteTransactions,
})

export const navigateToPaymentTransferReview = (
  type: TokenTransactionType,
  timestamp: number,
  confirmationProps: TransferConfirmationCardProps
) => {
  navigate(Screens.TransactionReview, {
    reviewProps: {
      type,
      timestamp,
    },
    confirmationProps,
  })
}

export const navigateToExchangeReview = (
  timestamp: number,
  confirmationProps: ExchangeConfirmationCardProps
) => {
  navigate(Screens.TransactionReview, {
    reviewProps: {
      type: TokenTransactionType.Exchange,
      timestamp,
    },
    confirmationProps,
  })
}

export const navigateToRequestedPaymentReview = (transactionData: TransactionDataInput) => {
  navigate(Screens.SendConfirmationLegacy, { transactionData, origin: SendOrigin.AppRequestFlow })
}
