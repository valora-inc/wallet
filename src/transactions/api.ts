import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { type LocalCurrencyCode } from 'src/localCurrency/consts'
import { getMultichainFeatures } from 'src/statsig'
import { FEED_V2_INCLUDE_TYPES, type PageInfo, type TokenTransaction } from 'src/transactions/types'
import networkConfig from 'src/web3/networkConfig'

export type TransactionFeedV2Response = {
  transactions: TokenTransaction[]
  pageInfo: PageInfo
}

const baseQuery = fetchBaseQuery({
  baseUrl: networkConfig.getWalletTransactionsUrl,
  headers: { Accept: 'application/json' },
})

export const transactionFeedV2Api = createApi({
  reducerPath: 'transactionFeedV2Api',
  baseQuery,
  endpoints: (builder) => ({
    transactionFeedV2: builder.query<
      TransactionFeedV2Response,
      {
        address: string
        localCurrencyCode: LocalCurrencyCode
        endCursor: PageInfo['endCursor'] | undefined
      }
    >({
      query: ({ address, localCurrencyCode, endCursor: afterCursor }) => {
        return {
          url: '',
          params: {
            address,
            networkIds: getMultichainFeatures().showTransfers.join(','),
            includeTypes: FEED_V2_INCLUDE_TYPES.join(','),
            localCurrencyCode,
            ...(afterCursor && { afterCursor }),
          },
        }
      },
      keepUnusedDataFor: 60, // 1 min
      transformErrorResponse: (error, meta) => {
        if (meta) {
          const params = new URL(meta.request.url).searchParams
          return {
            ...error,
            // only requests for next pages have the afterCursor search param
            hasAfterCursor: params.get('afterCursor'),
          }
        }

        return error
      },
    }),
  }),
})

export const { useTransactionFeedV2Query } = transactionFeedV2Api
