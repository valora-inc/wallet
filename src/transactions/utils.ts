import { TokenTransactionType } from 'src/apollo/types'
import { ExchangeConfirmationCardProps } from 'src/exchange/ExchangeConfirmationCard'
import i18n from 'src/i18n'
import { AddressToDisplayNameType } from 'src/identity/reducer'
import { TransferConfirmationCardProps } from 'src/transactions/TransferConfirmationCard'
import { Currency } from 'src/utils/currencies'
import { formatFeedSectionTitle, timeDeltaInDays } from 'src/utils/time'

// Groupings:
// Recent -> Last 7 days.
// [Current month] - "July" -> Captures transactions from the current month that aren’t captured in Recent.
// [Previous months] - "June" -> Captures transactions by month.
// [Months over a year ago] — "July 2019" -> Same as above, but with year appended.
// Sections are hidden if they have no items.
export function groupFeedItemsInSections<T extends { timestamp: number }>(items: T[]) {
  const sectionsMap: {
    [key: string]: {
      data: T[]
      daysSinceTransaction: number
    }
  } = {}

  items.reduce((sections, item) => {
    const daysSinceTransaction = timeDeltaInDays(Date.now(), item.timestamp)
    const key =
      daysSinceTransaction <= 7
        ? i18n.t('feedSectionHeaderRecent')
        : formatFeedSectionTitle(item.timestamp, i18n)
    sections[key] = sections[key] || {
      daysSinceTransaction,
      data: [],
    }
    sections[key].data.push(item)
    return sections
  }, sectionsMap)

  return Object.entries(sectionsMap)
    .sort((a, b) => a[1].daysSinceTransaction - b[1].daysSinceTransaction)
    .map(([key, value]) => ({
      title: key,
      data: value.data,
    }))
}

export const exchangeReviewHeader = (confirmationProps: ExchangeConfirmationCardProps) => {
  const { makerAmount } = confirmationProps
  const isSold = makerAmount.currencyCode === Currency.Celo
  return isSold ? i18n.t('soldGold') : i18n.t('purchasedGold')
}

export const transferReviewHeader = (
  type: TokenTransactionType,
  confirmationProps: TransferConfirmationCardProps,
  addressToDisplayName: AddressToDisplayNameType,
  rewardsSenders: string[]
) => {
  let headerText = ''
  const isCeloTx = confirmationProps.amount.currencyCode === Currency.Celo
  switch (type) {
    case TokenTransactionType.Sent:
      headerText = i18n.t(isCeloTx ? 'transactionHeaderWithdrewCelo' : 'transactionHeaderSent')
      break
    case TokenTransactionType.EscrowSent:
      headerText = i18n.t('transactionHeaderEscrowSent')
      break
    case TokenTransactionType.Received:
      const address = confirmationProps.address ?? ''
      if (rewardsSenders.includes(address) || addressToDisplayName[address]?.isCeloRewardSender) {
        headerText = i18n.t('transactionHeaderCeloReward')
      } else {
        headerText = isCeloTx
          ? i18n.t('transactionHeaderCeloDeposit')
          : i18n.t('transactionHeaderReceived')
      }
      break
    case TokenTransactionType.EscrowReceived:
      headerText = i18n.t('transactionHeaderEscrowReceived')
      break
    case TokenTransactionType.VerificationFee:
      headerText = i18n.t('transactionHeaderVerificationFee')
      break
    case TokenTransactionType.Faucet:
      headerText = i18n.t('transactionHeaderFaucet')
      break
    case TokenTransactionType.InviteSent:
      headerText = i18n.t('transactionHeaderInviteSent')
      break
    case TokenTransactionType.InviteReceived:
      headerText = i18n.t('transactionHeaderInviteReceived')
      break
    case TokenTransactionType.NetworkFee:
      headerText = i18n.t('transactionHeaderNetworkFee')
      break
  }

  return headerText
}
