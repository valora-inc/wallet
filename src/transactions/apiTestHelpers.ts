import type { EnhancedStore, Middleware, Reducer, UnknownAction } from '@reduxjs/toolkit'
import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { ApiReducersKeys } from 'src/redux/apiReducersList'
import { RootState } from 'src/redux/reducers'
import { getMockStoreData, RecursivePartial } from 'test/utils'

/**
 * This function is taken from the Redux team. It creates a testable store that is compatible
 * with RTK-Query. It is slightly modified to also include the preloaded state.
 * https://github.com/reduxjs/redux-toolkit/blob/e7540a5594b0d880037f2ff41a83a32c629d3117/packages/toolkit/src/tests/utils/helpers.tsx#L186
 *
 * For more info on why this is needed and how it works - here's an article that answers some of the questions:
 * https://medium.com/@johnmcdowell0801/testing-rtk-query-with-jest-cdfa5aaf3dc1
 */
export function setupApiStore<
  A extends {
    reducer: Reducer<any, any>
    reducerPath: string
    middleware: Middleware
  },
  Preloaded extends RecursivePartial<Omit<RootState, ApiReducersKeys>>,
  R extends Record<string, Reducer<any, any>> = Record<never, never>,
>(api: A, preloadedState: Preloaded, extraReducers?: R) {
  // destructure _persist from the preloaded state as it is an unexpected key in
  // the mocked store and logs an error during test execution
  const { _persist, ...mockStoreData } = getMockStoreData(preloadedState)
  const getStore = () =>
    configureStore({
      preloadedState: mockStoreData,
      reducer: combineReducers({
        [api.reducerPath]: api.reducer,
        ...extraReducers,
      }),
      middleware: (getDefaultMiddleware) => {
        return getDefaultMiddleware({
          serializableCheck: false,
          immutableCheck: false,
        }).concat(api.middleware)
      },
    })

  type Store = { api: ReturnType<A['reducer']> } & { [K in keyof R]: ReturnType<R[K]> }
  type StoreType = EnhancedStore<
    Store,
    UnknownAction,
    ReturnType<typeof getStore> extends EnhancedStore<any, any, infer M> ? M : never
  >

  const initialStore = getStore() as StoreType
  const refObj = { api, store: initialStore }
  const store = getStore() as StoreType
  refObj.store = store

  return refObj
}
