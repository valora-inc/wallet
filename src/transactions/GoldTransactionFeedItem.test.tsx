// @ts-ignore
import { toBeDisabled } from '@testing-library/jest-native'
import { render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { TokenTransactionType } from 'src/apollo/types'
import GoldTransactionFeedItem from 'src/transactions/GoldTransactionFeedItem'
import { TransactionStatus } from 'src/transactions/types'
import { createMockStore, getElementText, getMockI18nProps } from 'test/utils'

expect.extend({ toBeDisabled })

const localAmount = {
  value: '1.23',
  exchangeRate: '0.555',
  currencyCode: 'EUR',
}

describe('GoldTransactionFeedItem', () => {
  let dateNowSpy: any
  beforeAll(() => {
    // Lock Time
    dateNowSpy = jest.spyOn(Date, 'now').mockImplementation(() => 1487076708000)
    // set the offset to ALWAYS be Pacific for these tests regardless of where they are run
    // see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTimezoneOffset
    jest.spyOn(Date.prototype, 'getTimezoneOffset').mockImplementation(() => 420)
  })

  afterAll(() => {
    // Unlock Time
    dateNowSpy.mockRestore()
  })

  it('renders correctly', () => {
    const tree = render(
      <Provider store={createMockStore({})}>
        <GoldTransactionFeedItem
          status={TransactionStatus.Complete}
          __typename="TokenExchange"
          type={TokenTransactionType.Exchange}
          hash={'0x'}
          amount={{ value: '-1', currencyCode: 'cUSD', localAmount }}
          makerAmount={{ value: '1', currencyCode: 'cUSD', localAmount }}
          takerAmount={{ value: '10', currencyCode: 'cGLD', localAmount }}
          timestamp={1}
          {...getMockI18nProps()}
        />
      </Provider>
    )
    expect(tree).toMatchSnapshot()

    expect(getElementText(tree.getByTestId('GoldTransactionFeedItemRate/value'))).toEqual('€0.55')
    expect(getElementText(tree.getByTestId('GoldTransactionFeedItemAmount/value'))).toEqual('€1.23')
  })

  it('renders correctly when local amount is null', () => {
    const tree = render(
      <Provider store={createMockStore({})}>
        <GoldTransactionFeedItem
          status={TransactionStatus.Complete}
          __typename="TokenExchange"
          type={TokenTransactionType.Exchange}
          hash={'0x'}
          amount={{ value: '-1', currencyCode: 'cUSD', localAmount: null }}
          makerAmount={{ value: '1', currencyCode: 'cUSD', localAmount: null }}
          takerAmount={{ value: '10', currencyCode: 'cGLD', localAmount: null }}
          timestamp={1}
          {...getMockI18nProps()}
        />
      </Provider>
    )
    expect(tree).toMatchSnapshot()

    // This is a degraded mode, when we can't get the exchange rate from the blockchain-api, better than nothing
    expect(getElementText(tree.getByTestId('GoldTransactionFeedItemRate/value'))).toEqual('-')
    expect(getElementText(tree.getByTestId('GoldTransactionFeedItemAmount/value'))).toEqual(
      '-₱1.33'
    )
  })

  it('renders correctly when local amount is null and no local exchange rate is set', () => {
    const tree = render(
      <Provider store={createMockStore({ localCurrency: { exchangeRates: {} } })}>
        <GoldTransactionFeedItem
          status={TransactionStatus.Complete}
          __typename="TokenExchange"
          type={TokenTransactionType.Exchange}
          hash={'0x'}
          amount={{ value: '-1', currencyCode: 'cUSD', localAmount: null }}
          makerAmount={{ value: '1', currencyCode: 'cUSD', localAmount: null }}
          takerAmount={{ value: '10', currencyCode: 'cGLD', localAmount: null }}
          timestamp={1}
          {...getMockI18nProps()}
        />
      </Provider>
    )
    expect(tree).toMatchSnapshot()

    // This is an even more degraded mode, when we can't get the exchange rate from the blockchain-api or locally, better than nothing
    expect(getElementText(tree.getByTestId('GoldTransactionFeedItemRate/value'))).toEqual('-')
    expect(getElementText(tree.getByTestId('GoldTransactionFeedItemAmount/value'))).toEqual('-')
  })

  it('tap disabled while pending', () => {
    const { getByTestId } = render(
      <Provider store={createMockStore({})}>
        <GoldTransactionFeedItem
          status={TransactionStatus.Pending}
          __typename="TokenExchange"
          type={TokenTransactionType.Exchange}
          hash={'0x'}
          amount={{ value: '-1', currencyCode: 'cUSD', localAmount }}
          makerAmount={{ value: '1', currencyCode: 'cUSD', localAmount }}
          takerAmount={{ value: '10', currencyCode: 'cGLD', localAmount }}
          timestamp={1}
          {...getMockI18nProps()}
        />
      </Provider>
    )
    expect(getByTestId('GoldTransactionFeedItem')).toBeDisabled()
  })
})
