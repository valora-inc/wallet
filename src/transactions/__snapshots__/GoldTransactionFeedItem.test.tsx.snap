// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoldTransactionFeedItem renders correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  nativeBackgroundAndroid={
    Object {
      "attribute": "selectableItemBackground",
      "rippleRadius": undefined,
      "type": "ThemeAttrAndroid",
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="GoldTransactionFeedItem"
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
        "padding": 16,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "space-between",
          "paddingBottom": 2,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          feedItemGoldPurchased
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "flexWrap": "wrap",
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
           @ 
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "flexWrap": "wrap",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
              },
            ]
          }
          testID="GoldTransactionFeedItemRate/value"
        >
          
          €
          0.55
        </Text>
      </View>
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
              },
            ]
          }
          testID="GoldTransactionFeedItemAmount/value"
        >
          
          €
          1.23
        </Text>
      </View>
    </View>
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#9CA4A9",
            "fontFamily": "Inter-Regular",
            "fontSize": 14,
            "lineHeight": 18,
          }
        }
      >
        Dec 31 at 5:00 PM
      </Text>
    </View>
  </View>
</View>
`;

exports[`GoldTransactionFeedItem renders correctly when local amount is null 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  nativeBackgroundAndroid={
    Object {
      "attribute": "selectableItemBackground",
      "rippleRadius": undefined,
      "type": "ThemeAttrAndroid",
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="GoldTransactionFeedItem"
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
        "padding": 16,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "space-between",
          "paddingBottom": 2,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          feedItemGoldPurchased
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "flexWrap": "wrap",
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
           @ 
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "flexWrap": "wrap",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
              },
            ]
          }
          testID="GoldTransactionFeedItemRate/value"
        >
          
          -
        </Text>
      </View>
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
              },
            ]
          }
          testID="GoldTransactionFeedItemAmount/value"
        >
          -
          ₱
          1.33
        </Text>
      </View>
    </View>
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#9CA4A9",
            "fontFamily": "Inter-Regular",
            "fontSize": 14,
            "lineHeight": 18,
          }
        }
      >
        Dec 31 at 5:00 PM
      </Text>
    </View>
  </View>
</View>
`;

exports[`GoldTransactionFeedItem renders correctly when local amount is null and no local exchange rate is set 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  nativeBackgroundAndroid={
    Object {
      "attribute": "selectableItemBackground",
      "rippleRadius": undefined,
      "type": "ThemeAttrAndroid",
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="GoldTransactionFeedItem"
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
        "padding": 16,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "space-between",
          "paddingBottom": 2,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          feedItemGoldPurchased
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "flexWrap": "wrap",
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
           @ 
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "flexWrap": "wrap",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
              },
            ]
          }
          testID="GoldTransactionFeedItemRate/value"
        >
          
          -
        </Text>
      </View>
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
              },
            ]
          }
          testID="GoldTransactionFeedItemAmount/value"
        >
          
          -
        </Text>
      </View>
    </View>
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#9CA4A9",
            "fontFamily": "Inter-Regular",
            "fontSize": 14,
            "lineHeight": 18,
          }
        }
      >
        Dec 31 at 5:00 PM
      </Text>
    </View>
  </View>
</View>
`;
